### ISSUE TITLE
Introspection fails to generate `@relation` for many-to-many relation in MySQL

### ISSUE THREAD

**Issue by Akxe** [prisma/prisma#14668]
Newer reproduction here: https://github.com/prisma/prisma/issues/14668#issuecomment-1755468605

---

### Bug description

Introspection (`db pull`) fails to generate `@relation` for relations between tables, that are using compound keys.

### How to reproduce

1. Create schema from SQL bellow
2. run `npx prisma db pull`
3. See error in relation between tables

### Expected behavior

The resulting schema should create `@relation` with the correct fields.

Generated:
```
@relation(fields: [offerID], references: [offerID], onDelete: Cascade, onUpdate: NoAction)
```

Correct:
```
@relation(fields: [offerID, periodID, carrierID], references: [offerID, periodID, carrierID], onDelete: Cascade, onUpdate: NoAction)
```

### Prisma information

```sql
CREATE TABLE `nabzak` (
  `offerID` smallint unsigned NOT NULL,
  `periodID` smallint unsigned NOT NULL,
  `carrierID` int NOT NULL,
  PRIMARY KEY (`offerID`, `periodID`, `carrierID`),
  CONSTRAINT `nabzak_zaknos` FOREIGN KEY (`offerID`, `periodID`, `carrierID`) REFERENCES `CarrierInPeriod` (`offerID`, `periodID`, `carrierID`) ON DELETE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=cp1250 COLLATE=cp1250_czech_cs;

CREATE TABLE `CarrierInPeriod` (
  `offerID` smallint unsigned NOT NULL,
  `periodID` smallint unsigned NOT NULL,
  `carrierID` int NOT NULL,
  PRIMARY KEY (`offerID`, `periodID`, `carrierID`)
) ENGINE=InnoDB DEFAULT CHARSET=cp1250 COLLATE=cp1250_czech_cs;
```

### Environment & setup

- OS: Windows
- Database: MySQL
- Node.js version: v16.13.2

### Prisma Version

```
prisma                  : 4.1.1
@prisma/client          : 4.1.1
Current platform        : windows
Query Engine (Node-API) : libquery-engine 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Migration Engine        : migration-engine-cli 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules\@prisma\engines\migration-engine-windows.exe)
Introspection Engine    : introspection-core 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules\@prisma\engines\introspection-engine-windows.exe)
Format Binary           : prisma-fmt 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules\@prisma\engines\prisma-fmt-windows.exe)
Default Engines Hash    : 8d8414deb360336e4698a65aa45a1fbaf1ce13d8
Studio                  : 0.469.0
Preview Features        : interactiveTransactions
```

Extracted from https://github.com/prisma/prisma/issues/14648#issuecomment-1205224365

**Comment by janpio** [comment]
Hm, I ran your provided SQL in a new database and then `db pull`:

```prisma
model CarrierInPeriod {
  offerID   Int     @db.UnsignedSmallInt
  periodID  Int     @db.UnsignedSmallInt
  carrierID Int
  nabzak    nabzak?

  @@id([offerID, periodID, carrierID])
}

model nabzak {
  offerID         Int             @db.UnsignedSmallInt
  periodID        Int             @db.UnsignedSmallInt
  carrierID       Int
  CarrierInPeriod CarrierInPeriod @relation(fields: [offerID, periodID, carrierID], references: [offerID, periodID, carrierID], onDelete: Cascade, onUpdate: Restrict, map: "nabzak_zaknos")

  @@id([offerID, periodID, carrierID])
}
```

That does look correct to me. 

Is there anything else different between my schema and yours?

**Comment by Akxe** [comment]
I did not think it would change much, but my schema has a small difference; the order of primary index fields. 

```
@@id([offerID, carrierID, periodID])
```

**Comment by janpio** [comment]
In which of the models? Both?

**Comment by Akxe** [comment]
> In which of the models? Both?

Just in `nabzak`.

**Comment by janpio** [comment]
Hm ok, but we are talking about `db pull` here. If the SQL has the fields in that order, Introspection will also create the `@@id` with the fields in that order. Is there also a difference in the SQL that you provided?

**Comment by Akxe** [comment]
@janpio This is what `db pull` created among many other tables: 

```mysql
CREATE TABLE `nabzak` (
  `idZak` smallint unsigned NOT NULL,
  `idObdobi` smallint unsigned NOT NULL,
  `ncis` int NOT NULL,
  `obsaz` char(2) CHARACTER SET cp1250 COLLATE cp1250_czech_cs DEFAULT NULL,
  `X` tinyint DEFAULT NULL,
  PRIMARY KEY (`idZak`,`ncis`,`idObdobi`),
  KEY `nabzak_zaknos` (`idZak`,`idObdobi`,`ncis`),
  KEY `ncis` (`ncis`) /*!80000 INVISIBLE */,
  KEY `idZak` (`idZak`),
  CONSTRAINT `nabzak_zaknos` FOREIGN KEY (`idZak`, `idObdobi`, `ncis`) REFERENCES `zaknos` (`idZak`, `idObdobi`, `ncis`) ON DELETE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=cp1250 COLLATE=cp1250_czech_cs;


CREATE TABLE `zaknos` (
  `idZak` smallint unsigned NOT NULL,
  `idObdobi` smallint unsigned NOT NULL,
  `ncis` int NOT NULL,
  `X` tinyint(1) NOT NULL DEFAULT '0',
  `XZ` tinyint(1) NOT NULL DEFAULT '0',
  `Aud` tinyint(1) NOT NULL DEFAULT '0',
  `Obsaz` char(2) NOT NULL DEFAULT '',
  `Motiv` smallint unsigned DEFAULT '1',
  `PCena` mediumint DEFAULT NULL,
  `TCena` smallint unsigned DEFAULT NULL,
  `Pocet` smallint NOT NULL DEFAULT '1',
  PRIMARY KEY (`idZak`,`idObdobi`,`ncis`),
  KEY `IDZak & IDObdobi` (`idZak`,`idObdobi`) /*!80000 INVISIBLE */,
  KEY `IDZak & ncis` (`idZak`,`ncis`),
  KEY `zaknos_motiv` (`Motiv`),
  CONSTRAINT `zaknos_motiv` FOREIGN KEY (`Motiv`) REFERENCES `motivy` (`Motiv`) ON DELETE SET NULL,
  CONSTRAINT `zaknos_zakobdobi` FOREIGN KEY (`idZak`, `idObdobi`) REFERENCES `zakobdobi` (`idZak`, `idObdobi`) ON DELETE CASCADE,
  CONSTRAINT `zaknos_zakp` FOREIGN KEY (`idZak`, `ncis`) REFERENCES `zakazkyp` (`idZak`, `NCis`) ON DELETE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=cp1250;

```

I tried to reduce it to a minimum but probably left some things out in the process. I think that removing `CONSTRAINT`s from `zaknos` should be enough to make it valid database while keeping the error in.

**Comment by janpio** [comment]
Hm, with this SQL I can only reproduce the problem from https://github.com/prisma/prisma/issues/14648 again - the `@relation.fields` is dismissed because the `@@id` has a different order of fields. I can not reproduce a `@relation` with only one field instead of the expected 3.

**Comment by Akxe** [comment]
@janpio Delete your previous model, it can connect the dots when the `schema.prisma` already existed

**Comment by janpio** [comment]
I did:

```sql
CREATE TABLE `nabzak` (
  `idZak` smallint unsigned NOT NULL,
  `idObdobi` smallint unsigned NOT NULL,
  `ncis` int NOT NULL,
  PRIMARY KEY (`idZak`,`ncis`,`idObdobi`),
  KEY `nabzak_zaknos` (`idZak`,`idObdobi`,`ncis`),
  CONSTRAINT `nabzak_zaknos` FOREIGN KEY (`idZak`, `idObdobi`, `ncis`) REFERENCES `zaknos` (`idZak`, `idObdobi`, `ncis`) ON DELETE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=cp1250 COLLATE=cp1250_czech_cs;


CREATE TABLE `zaknos` (
  `idZak` smallint unsigned NOT NULL,
  `idObdobi` smallint unsigned NOT NULL,
  `ncis` int NOT NULL,
  PRIMARY KEY (`idZak`,`idObdobi`,`ncis`)
) ENGINE=InnoDB DEFAULT CHARSET=cp1250;

```
leads to
```prisma
model nabzak {
  idZak    Int    @db.UnsignedSmallInt
  idObdobi Int    @db.UnsignedSmallInt
  ncis     Int
  zaknos   zaknos @relation(fields: [idZak, idObdobi, ncis], references: [idZak, idObdobi, ncis], onDelete: Cascade, onUpdate: Restrict, map: "nabzak_zaknos")

  @@id([idZak, ncis, idObdobi])
  @@index([idZak, idObdobi, ncis], map: "nabzak_zaknos")
}

model zaknos {
  idZak    Int     @db.UnsignedSmallInt
  idObdobi Int     @db.UnsignedSmallInt
  ncis     Int
  nabzak   nabzak?

  @@id([idZak, idObdobi, ncis])
}
```

**Comment by Akxe** [comment]
Maybe the error got created long, long back, but the VSCode plugin did not recognize it as an error till that point. The last time I checked, the wrong order was present under 3.13 (the version I was before 4.1 and 4.2). It just probably did not get fixed but got reported as an error.

To be closed?

**Comment by janpio** [comment]
If you currently can not reproduce how you got a `@relation` with too little fields from Introspection, then I would say yes. As soon as you have some SQL that shows the same problem, I am more than happy to reopen this here of course. If this happens, it needs to be fixed.

**Comment by Akxe** [comment]
Hi @janpio, I have been busy...

I have made a 4 table repo. https://github.com/Akxe/prisma-n-to-n-bug/tree/master

Instruction:
1. `git clone https://github.com/Akxe/prisma-n-to-n-bug/tree/master`
2. `npm i`
3. use `initialize.sql` to create the 4 tables required
4. `npx prisma db pull -- --force`

In the [`prisma.schema` on line `31`](https://github.com/Akxe/prisma-n-to-n-bug/blob/13a4b1dec832164e6b961640d9ec010f620ba029/prisma/schema.prisma#L31) there is an error; the `CarrierInOffer` should be `CarrierInOffer[]`.

To explain why:
* One offer has multiple carriers
* One carrier can have multiple documentation, given they are on a different date
* The reverse from documentation to `CarrierOfOffer` does not point to a unique entry, but it is not necessarily a problem. I do not want to use the relation in the reversed direction.

**Comment by janpio** [comment]
Well, that doesn't happen too often that 1+ years later a working reproduction pops up - amazing job @Akxe.

I can indeed reproduce this with the repository you provided:
```
$ npx prisma validate
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma

Error: Prisma schema validation - (validate wasm)
Error code: P1012
error: Error parsing attribute "@relation": The argument `references` must refer to a unique criterion in the related model. Consider adding an `@unique` attribute to the field `carrierID` in the model `CarrierInOffer`.
  -->  schema.prisma:31
   | 
30 |   carrierID      Int            @db.UnsignedInt
31 |   CarrierInOffer CarrierInOffer @relation(fields: [carrierID], references: [carrierID], onDelete: Cascade, map: "documentationOfCarrier_offer")
32 | 
   | 

Validation Error Count: 1
[Context: validate]

Prisma CLI Version : 5.4.2
```

**Comment by janpio** [comment]
Although, let's dig in here a bit: Is `DocumentationOfCarrier` additional information for a specific date for the `Carrier`, or the `CarrierInOffer` (`Carrier` + `Offer`)? Right now it is `CarrierInOffer`, and that does not fully match your description I think.

This would be a more minimal reproduction:
```sql
CREATE TABLE `A` (
	`id1` int unsigned NOT NULL,
	`id2` int unsigned NOT NULL,
	PRIMARY KEY (`id1`, `id2`)
);

CREATE TABLE `B` (
	`date` date NOT NULL,
	`id1` int unsigned NOT NULL,
	PRIMARY KEY (`date`, `id1`),
	INDEX id1 (`id1`),
	CONSTRAINT `B_A` FOREIGN KEY (`id1`) REFERENCES `A` (`id1`) ON DELETE CASCADE ON UPDATE CASCADE
);
```

Leads to this Prisma schema:
```prisma
model A {
  id1 Int @db.UnsignedInt
  id2 Int @db.UnsignedInt
  B   B[]

  @@id([id1, id2])
}

model B {
  date DateTime @db.Date
  id1  Int      @db.UnsignedInt
  A    A        @relation(fields: [id1], references: [id1], onDelete: Cascade, map: "B_A")

  @@id([date, id1])
  @@index([id1], map: "id1")
}
```

Leads to this validation error:
```
$ npx prisma validate
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma

Error: Prisma schema validation - (validate wasm)
Error code: P1012
error: Error parsing attribute "@relation": The argument `references` must refer to a unique criterion in the related model. Consider adding an `@unique` attribute to the field `id1` in the model `A`.
  -->  schema.prisma:21
   | 
20 |   id1  Int      @db.UnsignedInt
21 |   A    A        @relation(fields: [id1], references: [id1], onDelete: Cascade, map: "B_A")
22 | 
   | 

Validation Error Count: 1
[Context: validate]

Prisma CLI Version : 5.4.2
```

**Comment by janpio** [comment]
With this minimal SQL the problems becomes clear:

When I convert the same SQL to PostgreSQL, I get this:

```sql
CREATE TABLE "A" (
    "id1" INTEGER NOT NULL,
    "id2" INTEGER NOT NULL,
    CONSTRAINT "A_pkey" PRIMARY KEY ("id1","id2")
);

CREATE TABLE "B" (
    "date" DATE NOT NULL,
    "id1" INTEGER NOT NULL,
    CONSTRAINT "B_pkey" PRIMARY KEY ("date","id1")
);

CREATE INDEX "id1" ON "B"("id1");
ALTER TABLE "B" ADD CONSTRAINT "B_A" FOREIGN KEY ("id1") REFERENCES "A"("id1") ON DELETE CASCADE ON UPDATE CASCADE;
```

But when I run it, PostgreSQL says "Nope":
```
Error: ERROR: there is no unique constraint matching given keys for referenced table "A"
```

The value in `references` is supposed to be unique and distinct.
When the column is only part of a multi column foreign key, that is not the case - so the foreign key should not be allowed.

Why does MySQL allow this!?

**Comment by Akxe** [comment]
You are right I kept those 2 additional tables for context.

In our database this exact problem occurs 3 times. Maybe more if I would fill all foreign key constraints.

My biggest problem is, that this MySQL produces invalid schema. At least commenting this out would do some benefit. However if I could use the relation to join these tables it would help me tremendously.

To work around this. I do edit the schema to delete the invalid generated relations. And in code, I do have to do multiple queries...

**Comment by Akxe** [comment]
> Although, let's dig in here a bit: Is `DocumentationOfCarrier` additional information for a specific date for the `Carrier`, or the `CarrierInOffer` (`Carrier` + `Offer`)? Right now it is `CarrierInOffer`, and that does not fully match your description I think.

In the most recent occurrence, I do have a table `CarrierInPeriod` that has the primary key of `offerID, carrierID, periodID`. This table is related to `Audit` with primary key `carrierID, periodID`. The reason why the audit has a smaller primary key is that the audit is true for the carrier regardless of the offer the carrier is currently in. It is either good, bad or needs improvement.

**Comment by Akxe** [comment]
> But when I run it, PostgreSQL says "Nope":
> 
> ```
> Error: ERROR: there is no unique constraint matching given keys for referenced table "A"
> ```

With this SQL MySQL would fail too. But if you add an index (non-uniques suffice), it will work… could you test it too? Just to be sure…

**Comment by janpio** [comment]
Can you suggest which SQL I should test? I am not fully sure.

**Comment by Akxe** [comment]
@janpio Hi, I can change it back. Just for you if you want :D

SQL like this? Maybe it could change the result?

```sql
CREATE TABLE "A" (
    "id1" INTEGER NOT NULL,
    "id2" INTEGER NOT NULL,
    CONSTRAINT "A_pkey" PRIMARY KEY ("id1","id2")
);

CREATE TABLE "B" (
    "date" DATE NOT NULL,
    "id1" INTEGER NOT NULL,
    CONSTRAINT "B_pkey" PRIMARY KEY ("date","id1")
);

CREATE INDEX "id1a" ON "A"("id1");
CREATE INDEX "id1b" ON "B"("id1");
ALTER TABLE "B" ADD CONSTRAINT "B_A" FOREIGN KEY ("id1") REFERENCES "A"("id1") ON DELETE CASCADE ON UPDATE CASCADE;
```

**Comment by janpio** [comment]
(Please don't, the yellow A is from the old days. You are a real person now 🤣)

I don't understand what you mean with "With this SQL MySQL would fail too". Of course the PostgreSQL SQL fails on MySQL, but the equivalent MySQL SQL works fine.

https://github.com/prisma/prisma/issues/14668#issuecomment-1758723951 is MySQL dialect SQL that I ran on a MySQL database. It works.
And https://github.com/prisma/prisma/issues/14668#issuecomment-1758728899 is then basically the same SQL adapted for PostgreSQL only.

The problem here really is that MySQL seems to allow you to define a foreign key in a way that should not be possible as you are missing the unique constraint on `A` that should be there. And if you do that in MySQL, Prisma protects you from not allowing you to write a Prisma schema like that. 

The data in `B` presents as it would point to 1 entry in `A`, but in reality it can point to many entries in `A`.  That is not supposed to be possible.

**Comment by Akxe** [comment]
It might be unconventional, but for our database, we use it in about five places, 3 of which I need and have to work around...

Do you want examples when it is needed, or is this a thing that you will not look into, and I should only look for alternatives? If I was to look for alternatives, could you help me with creating `prisma.$extends(...)` if it is even possible?

**Comment by janpio** [comment]
Yes, I would be very interested in what data requires this setup. Optimally with the example we already have (because I semi understand that) - or even the abstracted A-B minimized example. For now, this seems like a modelling mistake to me that should better be solved with another table in between so you have a real m:n relation if you want to have m:n data.

I don't think there is any real workaround here. You could put the suggested unique index in the schema so it is happy, and then never migrate that change - so permanently delete it from any generated migration. That would satisfy the schema validation, and keep the database schema unchanged. (Prisma Client extensions can't really be helpful as long as you have an invalid schema)

**Comment by Akxe** [comment]
Usually, it goes like this: A `carrier` has some kind of property that is unique in time (`period`): documentation (text and photo), assigned store(s). This information is however true regardless of the `offer` this carrier is currently in. In 99% a `carrier` is only present in one `offer` at a time. What is problematic is that a `offer` has usually 12 `period`s, thus to have a `carrier` continue in the next year, a new `offer` has to be made. We however do need to easily query for the documentation of a `carrier` in previous period`s thus we need to even look into past `offer`s.

```graphql
model Audit {
  carrierID Int             @map("ncis")
  periodID  Int             @map("idObdobi") @db.UnsignedSmallInt

  ...

  @@id([carrierID, periodID])
}
model CarrierInPeriod {
  offerID             Int                         @map("idZak") @db.UnsignedSmallInt
  periodID            Int                         @map("idObdobi") @db.UnsignedSmallInt
  carrierID           Int                         @map("ncis")
  
  ...

  @@id([offerID, periodID, carrierID])
  @@index([offerID, periodID], map: "IDZak & IDObdobi")
}
```

Is it understandable?

-------

PS: I also have to work around an active old project that uses the same database and I cannot modify it freely. (That is why every column is translated from Czech to English in the mappings)

**Comment by janpio** [comment]
> Is it understandable?

Honestly, not really. (Your code talks about `carrier` and `offer`, but then shows an `Audit` - unclear to me why and I would need to go back digging at older examples and so on, for which it is too late right now 😆)

For me it looks like the problematic relation is desired to be able to show that multiple A are can be connected to multiple B. If that is the case, a simple foreign key between A and B is not the correct tool anymore when using Prisma. Prisma will expect that one of such a relationship is the `1` side and is unique, and the other the `n` side. This might accidentally work right now, but this is not designed to work in Prisma.

I honestly think your best way forward might be to "lie" to Prisma by defining a `@unique` but not migrating that on the database. Prisma should not deal with the uniqueness locally at all, and hence you should be able to use your setup.

**Comment by Akxe** [comment]
@janpio I tried to show the most straightforward example to you in the comment before. **Also, I had an error in the index in the example above**. 

Does Prisma not allow many to many relations for other databases? Why not enable this feature for MySQL too?

PS: I will try the unique index, but I fear unforeseen side effects.

**Comment by janpio** [comment]
> Does Prisma not allow many to many relations for other databases? Why not enable this feature for MySQL too?

Many to many relations are not possible between and with only 2 tables in relational databases with foreign keys. You always have to use a join tables, which creates two 1-n relationships (2 foreign keys in the join table pointing to the primary key or unique of the 2 main tables).

**Comment by Akxe** [comment]
> Many to many relations are not possible between and with only 2 tables, with foreign keys, in relational databases. You always have to use a join table, which creates two 1-n relationships (2 foreign keys in the join table pointing to the primary key or unique of the 2 main tables). That is the only way this can be modelled - and that is what Prisma supports.

I used to think that too, yet my dad, who did not go to IT school (because there was none at that time), did prove me wrong. Clearly, we have gotten N-M relation in MySQL, and it makes all logical sense to me. 

From what I can tell, most teachers in schools tell their students to use one single ID column for every table, but omitting it gives the user much more flexibility, even the one we found out.

**Comment by janpio** [comment]
That is a non-standard expansion of SQL that InnoDB allows for MySQL, that we at Prisma just do not support. It changes the fundamental guarantees most users expect from a relation created via a foreign key - in that one side has to be unique.

**Comment by Akxe** [comment]
> It changes the fundamental guarantees most users expect from a relation created via a foreign key - in that one side has to be unique.

The guarantee is there as long as the reference column (set) is a unique or a primary key.

> That is a non-standard expansion of SQL that InnoDB allows for MySQL

Could you test the simple `A` `B` table in MariaDB? I wonder, if this is a MySQL only or if MariaDB has this too. Maybe it is only Postgress that does not support this... I don't know.

**Comment by janpio** [comment]
> The guarantee is there as long as the reference column (set) is a unique or a primary key.

Yes, but that is not the case in your example or the minified version I abstracted from it, isn't it? The column is part of a primary key with 2 columns, which means the column by itself is _not_ unique. As soon as you make it unique, Prisma will be happy with the relation.

**Comment by Akxe** [comment]
The guarantee is there. The user only has to understand that he did not specify the given combination as unique and as such the relation can be many to many (depending on other side...)

The problem is only there is the user specify non unique compound index on referenced columns. Given how specific this is. Isn't this something that have to be made consciously?

**Comment by janpio** [comment]
Either the database guarantees uniqueness via a constraint, or we can not rely on it being unique.
If there is no PK or unique constraint on the column, we have no guarantee.

Or am I missing something?

**Comment by Akxe** [comment]
Well, I do not see the foreign key constraint as a guarantee of uniqueness. To me, uniqueness is only guaranteed by the primary key or a unique key.

This might be the difference in our point of view... what do you thing?

**Comment by janpio** [comment]
A foreign key does not give uniqueness. But a foreign key can only point to a column that has a uniqueness guarantee. Except in MySQL, where this is not required. Prisma does not support that, and will not, as our whole system of relations is built on the standard definition of foreign keys - which includes a guarantee that the referenced field is unique because of the requirement to have a unique constraint of some kind.

**Comment by Akxe** [comment]
> Except in MySQL, where this is not required.

Are you sure? I tested `SQLite` and it allowed non-unique foreign keys. I am not sure how many other DB engines exist...

**Comment by janpio** [comment]
Ok, then let me rephrase (without having checked that): 
PostgreSQL does not allow it, as it should not. Others databases don't, but we do not want to support that.

**Comment by Akxe** [comment]
> we do not want to support that as it removes guarantees that we like and need.

That I can accept. So, to finish this off. 

To enable many-to-many relations in Prisma (with MySQL), I have to add `@@unique([column1, ..., columnN])`? Does a `@ignore` exists for this? I would love to use this when pulling DB so I do not have to re-add the key every time.

**Comment by janpio** [comment]
> To enable many-to-many relations in Prisma (with MySQL), I have to add `@@unique([column1, ..., columnN])`? 

No, to enable many-to-many relations you need to actually use many to many: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations There are two ways, implicit (where Prisma creates a join table under the hood) or explicit, where you do that manually and then actually have two 1-n relations.

> Does a `@ignore` exists for this? I would love to use this when pulling DB so I do not have to re-add the key every time.

No, and it should not exist. Either you have an index or constraint, or you don't. If you have to fake one existing, then you need toa add that every time. I suggest creating a script that does so if you really need to go that route.

**Comment by Akxe** [comment]
Is there a way to extend Prisma query to join with another table? I would like to have only one query to retrieve all info.

**Comment by janpio** [comment]
No, not right now. If you see an optimization we could do, open a new issue and describe the schema, Prisma Client query and SQL query that could be optimized.

**Comment by Akxe** [comment]
> No, not right now. If you see an optimization we could do, open a new issue and describe the schema, Prisma Client query and SQL query that could be optimized.

Will do.

------

Back to the root of the problem. What should be done about the fact that `prisma db pull` does create an invalid schema? Should it just not create the relation? Or should it be created with a comment saying that is not supported?

**Comment by janpio** [comment]
I think the current state is fine, as this is an edge case. The database has a schema that Prisma can not support - so it gives the user an error message how they should fix it. If the user can add the unique constraint, they are fine. If they can not, they need to find a different solution like switching the relation to m:n.

If this pops up more often, it might be worth adding a special validation for the case of the missing unique. But we don't really know if a schema comes from Introspection, or was written by hand, when we validate - so that will be difficult to handle in a good way.

**Comment by Akxe** [comment]
TLDR for everyone with the same issue:

MySQL `InnoDB` (and a few others) allow the creation of many-to-many relations. Prisma does not support that, and adding it would prove very difficult. Prisma will not solve this as it is an edge case. DB with this relation will produce **invalid** schema when pulled (`prisma db pull`).

-----

To anyone reading this: If you have the issue, give this comment a like, and if more than 10 are present, write a message; maybe then something could be done about it.

**Comment by janpio** [comment]
To be precise, MySQL InnoDB allows foreign keys that point to a non-unique column. This makes the foreign key _behave_ a bit like a many-to-many relation, but [breaks other expected functionality](https://github.com/prisma/prisma/issues/14668#issuecomment-1763190928) of foreign keys like referential actions or `UPDATE`. We suggest you do not use this kind of foreign key, but either make the column unique or use a [real many-to-many relation](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations).
